Simplify management of distributed transactions.

hlinnaka opened the pull request:
We used to have a separate array of LocalDistributedXactData instances, and
a reference in PGPROC to its associated LocalDistributedXact. That's
unnecessarily complicated: we can store the LocalDistributedXact information
directly in the PGPROC entry, and get rid fo the auxiliary array and the
bookkeeping needed to manage that array.

This doesn't affect the backend-private cache of committed Xids that also
lives in cdblocaldistribxact.c

There's one performance thing worth calling out:
LocalDistribXact_LocalXidKnown() used to scan the linked list of
LocalDistributedXact entries, and it moved the found item to the head of the
list. That made repeated lookups of the same XID very cheap. This commit
gets rid of the linked list, and rewrites LocalDistribXact_LocalXidKnown()
to scan the proc array instead. The move-to-front optimization isn't possible
there, so repeatedly looking up the same XID is not any cheaper than looking
up random XIDs anymore. However, there is also a backend-private cache of
XIDs, which is checked before calling LocalDistribXact_LocalXidKnown(). You
could probably construct a test case to show the effect with in-progress XIDs
(or committed XIDs not visible to your snapshot; I'm not entirely sure how
this works), but I doubt that anyone will be seriously hurt by this in
practice.

hlinnaka commented:
Passed regression tests on my laptop. @ashwinstar, @asimrp, can you review this, please? Are there some extra test suites beside "make installcheck-good" that I could run this against?

ashwinstar commented:
Sure will take a look. This one GPDB-master-TINC-PgTwophase, suspends checkpoints and injects faults at interesting points from DTM perspective in different server states.

ashwinstar commented on src/backend/cdb/cdbtm.c:
```diff
@@ -3045,6 +3044,15 @@ generateGID(char *gid, DistributedTransactionId *gxid)
 }
 
+DistributedTransactionId
+getMaxDistributedXid(void)
```
How this function works for segments, as shmGIDSeq will only be incremented on master with generateGID.

hlinnaka replied:
Hmm. I thought you wouldn't need this in QEs, because distributed snapshots are always formed in the QD. But indeed, the gp_max_distributed_xid() function needs that, which is called at QD startup, to figure out which distributed XIDs are in use in the QEs, if the QEs have not been restarted together with the QD.

I'll fix that by making gp_max_distributed_xid() to the proc array for see the highest distributed XID in use, in the QEs. I'll add an commit to this PR for that shortly, stay tuned.

ashwinstar commented on src/backend/access/transam/twophase.c:
```diff
@@ -575,10 +576,10 @@ MarkAsPreparing(TransactionId xid,
+	gxact->proc.localDistribXactData.distribTimeStamp = distribTimeStamp;
+	gxact->proc.localDistribXactData.distribXid = distribXid;
+	gxact->proc.localDistribXactData.localXid = xid;
+	gxact->proc.localDistribXactData.state = LOCALDISTRIBXACT_STATE_PREPARED;
```
setting to LOCALDISTRIBXACT_STATE_PREPARED state here fine, we are still preparing ?

hlinnaka replied:
Ah, good catch. I was trying to preserve the old behaviour, and looked at the call from RecoverPreparedTransactions(). In that call, the state is initialized straight to LOCALDISTRIBXACT_STATE_PREPARED. But of course, for the call from PrepareTransaction(), it's not. Will fix.

ashwinstar commented:
LocalDistribXact was one of the code pieces skipped looking into but your PR made to look into it seriously :-) Unless I am really missing something not seeing need for maintaining this at all. Seeing it used for mapping LocalXid to DistributedXid through LocalDistribXact_LocalXidKnown() for snapshot checking. Apart from that only seeing it manipulated and set, but no decisions taken based on it. Like the states maintained doesn't seem to drive any decisions, correct ?

Getting to usage of LocalDistribXact_LocalXidKnown(), we will only call it when the localXid is found in committed state in clog, which means distributed_log is already written as it happens before clog. So, if we skip consulting this list and go to SLU / DISK, we can find the mapping to Dxid. Hence, its definitely not required for correctness, just used for optimization. Performing so much work just for optimization which is also for very small window on segments and little longer on master (between prepare and commit-prepared) after making it committed in clog and deleting the record from this list. With this lookup being more costly with traversing the full proclist, it seems more like should we completely get rid of all this code related to LocalDistribXact. Does that sound right ?

Thanks @asimrp for pointing out this. Need the max_distributed_transaction from segments, which should be maintained in shared_memory somewhere. Walking proc list would only give us current live sessions max_transaction_id and not max seen by this segment which is what we used to return when it was maintained in shared_memory, as with master reset most of the sessions would get disconnected from segments too.

asimrp commented on src/backend/cdb/cdbmaxdistributedxid.c:
```diff
@@ -24,10 +22,14 @@ PG_FUNCTION_INFO_V1(gp_max_distributed_xid);
 Datum
 gp_max_distributed_xid(PG_FUNCTION_ARGS __attribute__((unused)) )
```
Based on a quick reading of code, I am not sure if we need this function.  During recovery or normal start up, `tmShmemInit()` will be called.  This is when the timestamp component of distributed transaction ID is set.  As this is guaranteed to be unique across restarts of master / QD, why do we need to worry about setting `shmGIDSeq` to some value?  Rather, it should be reset to 0 always by `recoverTM()`.  No?

hlinnaka replied:
On 14/05/16 08:59, Asim RP wrote:
>> @@ -24,10 +22,14 @@ PG_FUNCTION_INFO_V1(gp_max_distributed_xid);
>>   Datum
>>   gp_max_distributed_xid(PG_FUNCTION_ARGS __attribute__((unused)) )
>
> Based on a quick reading of code, I am not sure if we need this
> function.  During recovery or normal start up, `tmShmemInit()` will
> be called.  This is when the timestamp component of distributed
> transaction ID is set.  As this is guaranteed to be unique across
> restarts of master / QD, why do we need to worry about setting
> `shmGIDSeq` to some value?  Rather, it should be reset to 0 always by
> `recoverTM()`.  No?

Oh, so that's what the timestamp is used for. Yeah, you're right, we 
don't need to guarantee that distributed XIDs are unique across 
restarts, if we always also compare the timestamp.

BTW, it seems to me that distributed XIDs are always compared with 
straight < and >. How do we handle distributed XID wraparound?

- Heikki

asimrp replied:
We don't handle distributed XID wraparound.  This is not hazardous because distributed XID is never recorded in a tuple.  Furthermore, if distributed XID wraps around, snapshot isolation will break only if we still have a looong looooong running transaction whose distributed XID was close to 1.  Possibility of this is negligible.  One way to handle wrap arounds is to generate a new timestamp after the 32-bit distributed XID integer wraps around.

hlinnaka replied:
On 18/05/16 12:46, Asim RP wrote:
> We don't handle distributed XID wraparound.

Ok.

> This is not hazardous because distributed XID is never recorded in a
> tuple. Furthermore, if distributed XID wraps around, snapshot
> isolation will break only if we still have a looong looooong running
> transaction whose distributed XID was close to 1. Possibility of this
> is negligible. One way to handle wrap arounds is to generate a new
> timestamp after the 32-bit distributed XID integer wraps around.

No, you can get incorrect query results when the wraparound happens, 
without a long running transaction, because the comparisons of a 
distributed XID and a distributed snapshot's XMIN and XMAX will produce 
wrong results at wraparound. 
DistributedSnapshotWithLocalMapping_CommittedTest() has e.g. this:

> 	/* Any xid < xmin is not in-progress */
> 	if (distribXid < header->xmin)
> 		return DISTRIBUTEDSNAPSHOT_COMMITTED_VISIBLE;

The snapshot's xmin might be e.g 2^32-1, if the snapshot was taken just 
before the wraparound, and distribXid might be a distributed XID that 
was assigned just after the wraparound, e.g. 5. In that case, distribXid 
is logically larger than the xmin, but that straight comparison above 
will incorrectly conclude that it's smaller, and that the transaction 
should be visible to you.

This is why you always have to use TransactionIdFollows() and friends 
when comparing XIDs.

- Heikki

asimrp replied:
That's right, wraparound of distributed xid indeed breaks visibility.  May be, that's the reason we simply prevent a wrap around in `generateGID()` like so:(!!)
```
    if (*shmGIDSeq >= LastDistributedTransactionId)
    {
        releaseTmLock();
        ereport(FATAL,
            (errmsg("reached limit of %u global transactions per start", LastDistributedTransactionId)));
    }
```

I think we can live with this for now, but at some point, we should handle wrap around of distributed XID more seriously.

hlinnaka replied:
On 18 May 2016 19:37:13 GMT-04:00, Asim RP <notifications@github.com> wrote:
>> @@ -24,10 +22,14 @@ PG_FUNCTION_INFO_V1(gp_max_distributed_xid);
>>  Datum
>>  gp_max_distributed_xid(PG_FUNCTION_ARGS __attribute__((unused)) )
>
>That's right, wraparound of distributed xid indeed breaks visibility. 
>May be, that's the reason we simply prevent a wrap around in
>`generateGID()` like so:(!!)
>```
>    if (*shmGIDSeq >= LastDistributedTransactionId)
>    {
>        releaseTmLock();
>        ereport(FATAL,
>(errmsg("reached limit of %u global transactions per start",
>LastDistributedTransactionId)));
>    }
>```

Ah, I see :)

>I think we can live with this for now, but at some point, we should
>handle wrap around of distributed XID more seriously.

Yep.

- Heikki

hlinnaka commented:
On 14/05/16 04:47, Ashwin Agrawal wrote:
> LocalDistribXact was one of the code pieces skipped looking into but
> your PR made to look into it seriously :-) Unless I am really missing
> something not seeing need for maintaining this at all. Seeing it used
> for mapping LocalXid to DistributedXid through
> LocalDistribXact_LocalXidKnown() for snapshot checking. Apart from
> that only seeing it manipulated and set, but no decisions taken based
> on it. Like the states maintained doesn't seem to drive any
> decisions, correct ?

A-ha, brilliant!

> Getting to usage of LocalDistribXact_LocalXidKnown(), we will only
> call it when the localXid is found in committed state in clog, which
> means distributed_log is already written as it happens before clog.
> So, if we skip consulting this list and go to SLU / DISK, we can find
> the mapping to Dxid. Hence, its definitely not required for
> correctness, just used for optimization. Performing so much work just
> for optimization which is also for very small window on segments and
> little longer on master (between prepare and commit-prepared) after
> making it committed in clog and deleting the record from this list.
> With this lookup being more costly with traversing the full proclist,
> it seems more like should we completely get rid of all this code
> related to LocalDistribXact. Does that sound right ?

Yeah, I think you're right. This comment in 
DistributedSnapshotWithLocalMapping_CommittedTest():

> /*
>  * Small window -- but check if our LocalDistribXact element says
>  * we are preparing this transaction.  If so, the CLOG knows the
>  * commit and had the visibility routine call us before we've finished
>  * updating our data structures.
>  */
> if (!LocalDistribXact_LocalXidKnown(
> 				localXid,
> 				dslm->header.distribTransactionTimeStamp,
> 				&distribXid))

suggests that the clog is updated before the distributedlog SLRU (that's 
what "our data structures" refers to, I believe). But looking at 
RecordTransactionCommit, the clog is updated first, as you pointed out.

Since PGPROC->localDistribXactData is never accessed by other backends, 
that information doesn't actually need to be in PGPROC at all. We could 
store that information simply as a global variable, in backend private 
memory. But I was hesitant to go that far, as the distributed XID seems 
potentially very useful for debugging purposes. I actually think we 
should have a view similar to, or part of, pg_prepared_xacts or 
pg_stat_activity, that shows the distributed XID of each in-progress 
transaction. But that's a different story. I did simplify the locking 
and the code that updates that, though, based on the assumption that 
it's not used by other backends.

I'll simplify that further, and update the PR. Thanks!

- Heikki

ashwinstar commented:
TransactionStateData already stores distribXid, which is used at logging to put "dx__" in logs, can add timestamp same place if helps for debugging (full gid)

ascherbaum-pivotal commented:
Any idea about the real world overhead when a scan of the proc array is required all the time, instead of a pre-sorted list in LocalDistributedXact?

hlinnaka commented:
On 16/05/16 22:15, Andreas Scherbaum wrote:
> Any idea about the real world overhead when a scan of the proc array
> is required all the time, instead of a pre-sorted list in
> LocalDistributedXact?

I haven't measured it, but I doubt it's noticeable. The lookups are 
cached in a backend-private hash table, so it only makes a difference 
for the first lookup of a particular XID in a session.

- Heikki

hlinnaka commented:
Rebased over master (unsurprisingly there were some conflicts with the 8.3 merge), and squashed and reordered the commits. Everyone, please have another look.

ashwinstar commented on src/backend/access/transam/xact.c:
```diff
@@ -3652,17 +3636,13 @@ CommitTransaction(void)
-	if (!LocalDistribXactRef_IsNil(&localDistribXactRef))
+	if (MyProc->localDistribXactData.state != LOCALDISTRIBXACT_STATE_NONE)
 	{
 		LWLockAcquire(ProcArrayLock, LW_EXCLUSIVE);
```
Do we need to acquire ProcArrayLock here ?

ashwinstar commented on src/include/storage/proc.h:
```diff
@@ -87,14 +87,19 @@ struct PGPROC
+	 * could be just a global variable in backend-private memory, but it seems
+	 * useful to have this information available for debugging purposes.
+	 */
+	LocalDistribXactData localDistribXactData;
```
Was earlier thinking we would get rid of this all together as states and all in this is not used anywhere, but yes maybe helpful for debugging so fine to have it.

ashwinstar commented:
👍  Looks good to me. Thank You.

hlinnaka commented:
Ok, pushed. Thanks for all the reviews!
